### ISSUE TITLE
Multi-node (6000+) polygon editing performance issue

### ISSUE THREAD

**Issue by cihaniko** [codeofsumit/leaflet.pm#366]
Hi,

I want to do multi-node (6000+ node) polygon editing. However, the browser runs slowly or crashes because all nodes appear in edit mode.

What can we do to eliminate this problem?

Is it possible to show only the nodes in Bound and edit them in this way? In this way, not all nodes are displayed and no performance problems.

Example jsfiddle: http://jsfiddle.net/wjo5enpq/5/   (Holds a little while loading. Please wait loading.)

I need help,
Thank you.

**Comment by themre** [comment]
it's probably better to group markers based on some distance... e.g. you show only one markers if distance between them is less than 5 pixels... if you zoom in, you show them. however this won't solve all problems, so it's probably also better to also show only the ones that are inside viewport.

**Comment by codeofsumit** [comment]
@themre interesting. But what is a marker-group for? You can't drag it, you can only click it to zoom in and reveal the individual markers. Just like marker-cluster. Why show them at all then?

**Comment by themre** [comment]
yeah, since it's a group, you need to show users that there are some many nodes there, they need to zoom in. because if you give 1 random node for editing, they will draw spikes because other nodes will be there. you can give user a decision and on hover offer 2 actions: zoom to nodes or move whole group of nodes that are behind that marker.

**Comment by themre** [comment]
now that we are talking about groups... it would probably a nice feature to add selection of nodes... e.g. hold Ctrl and drag to select nodes inside rectangle :)

**Comment by cihaniko** [comment]
> Uuuhhh çok havalı! Bunun gibi bir dava için bekliyorum.
> 
> Seninle aynı fikre sahiptim. Aşağıdaki mantıkla yalnızca sınırlı miktarda düzenleme noktasını göster:
> 
> * Fare imlecine yakın köşeler için yalnızca işaretçileri göster `distance < x`
> * Sadece geçerli görünümde bulunan işaretçileri göster (mobil için)
> * Maksimum `y`işaretçiyi göster (yakınlaştırmak için)
> 
> Ne düşünüyorsun?

**Comment by cihaniko** [comment]
@themre I thought about the first grouping. But it is very difficult to manage. For example, if we group them if the distance between them is less than 5 pixels, some nodes remain in the group even at the maximum zoom level. This may be undesirable.

@codeofsumit Therefore, it makes more sense to show the nodes within the limits of certain zoom levels in edit mode. The developer specifies the zoom level. At this level, a better performance can be obtained by showing the remaining nodes within the boundary.
How to display the mouse cursor when the zoom level is low For example, how will the cursor appear at zoom level 8 on a wide polygon?

**Comment by themre** [comment]
you can still disable grouping if there are less than e.g. 100 nodes. adding additional zoom levels seems a bit of an overkill. library should try to prevent browser crash due to too many nodes. so if a shape has more than 100 nodes, start grouping. otherwise leave as it is. at least you have some safety not to crash browser.

**Comment by codeofsumit** [comment]
I'm still not convinced of grouping. It has no benefit to the user IMHO. The user has to zoom in to edit anyway.

I'm not sure when I'm able to work on this but I'll make sure to share a prototype with you to review my proposed solution. It's easier to discuss then.

**Comment by hefler** [comment]
How about only showing *n* amount of nodes near the cursor?
However that might pose some interaction limitations under touch devices as there are no visual cues where your "cursor" is.

So a combination of zoom + proximity to the cursor might do the trick.

**Comment by codeofsumit** [comment]
@hefler exactly what I would like to do.
Still thinking about what to show when the user is zoomed out though. All n markers near the cursor would be on top of each other. While the user would have to zoom in to edit anyway it's not beneficial to show the markers either 🤔

**Comment by hefler** [comment]
@codeofsumit I think one could assume that if there are too many markers on a zoomed out view then it is the case of zooming in anyway. What about having a marker telling users that there are many other markers "under" and clicking on that marker + `fitBounds` kind of thing.
<pre>BEFORE                                        AFTER</pre>
![image](https://user-images.githubusercontent.com/195560/51027390-0df34a00-1591-11e9-90e6-b4cd202e62f2.png)

**Comment by codeofsumit** [comment]
this is a big one and I will definitely tackle it. As with other feature requests though, I will close the issue for now. If you want this implemented please upvote the original comment to help me prioritize feature requests 👍

**Comment by codeofsumit** [comment]
@lbusett yup everyone can implement this. I'm thinking of adding this as an option to the toolbar.
In this issue I would like to address the issue with a single huge polygon.

**Comment by ukhack** [comment]
Wow,it's cool

**Comment by Cran3kill** [comment]
@codeofsumit I've same problem with 1282 polygons (some has many more then 4 edges). Will you work on it anytime soon? I couldn't make it happen selected polygons on the other hand.

**Comment by nandojve** [comment]
Hello Guys,

I've facing this problem too. Today I discovered that my user will need edit 1 polygon with more than 100k nodes. This is the boundary of a small continent but in my country we have >820k nodes total to shape cover boundaries. What my group have been discussing is to give the zoom option to user too but we believe a better approach is the user have the capability to select a interest boundbox with 2 clicks and than enable that's nodes to edition. Once edition is ready user can save and repeat this process time to time.

This solution could be the same when we have lots of small polygons, lines or points. This feature should have parallel editing working capability using database because multiple users can edit same map in different regions. This allows multilayer edition too because user can edit layer by layer in that region one at time.

Well this are my 2 cents to this discussion.

**Comment by lobovmaxim** [comment]
@codeofsumit We have the same problem :(  Will you work on the problem?

**Comment by agh69** [comment]
@codeofsumit Wouldn't a similar tool as QGIS's "Reshape Feature" in the Advance Edit Toolbar be helpful? That way nodes won't have to be visible - just draw the new boarder where appropriate and update. At least I would benefit hugely from such a tool. Draging nodes one by one is painful. How to get there is unfortunately way beyond my understanding.

**Comment by codeofsumit** [comment]
@agh69 can you show a GIF/video of that feature?

**Comment by agh69** [comment]
@codeofsumit 
Reshape feature-tool in QGIS Advanced Digitizing panel:

![reshape_feature_qgis](https://user-images.githubusercontent.com/52559551/63919300-c3024600-ca3e-11e9-9411-f1856cb1577a.gif)

**Comment by codeofsumit** [comment]
![cut](https://user-images.githubusercontent.com/2399810/63919728-a0246180-ca3f-11e9-8441-0c81004ebdd7.gif)
 isn't the cutting feature sufficient?

**Comment by agh69** [comment]
@codeofsumit 
If you have two (or more) adjacent polygons the cutting feature will create a hole where you cut, whereas the reshape feature moves the boarder to the drawn line.
![cutting_feature_pm](https://user-images.githubusercontent.com/52559551/63920669-a9aec900-ca41-11e9-955b-f8fcdabef244.gif)

**Comment by codeofsumit** [comment]
Ok got it that is indeed a different feature and it raises some questions for me. But it is also not part of this issue. Please create a new issue/feature request for this with an explanation of how it works so we can discuss how this is doable there :-)
Thanks for raising it 🎉

**Comment by agh69** [comment]
Yes, sure. I wanted to point out that the QGIS reshape-tool does not show the nodes while in editing mode, and that might be part of a solution for the original question. On the other hand, cutting tool doesn't show them either. 
Anyway, thanks for all the fantastic contribution you do to the world!

**Comment by lobovmaxim** [comment]
Hi guys, I have fixed the problem (Multi-node (6000+) polygon editing performance) about 2 month ago and it works great.  The decision is show only N count of markers near cursor (optionaly). I thought about pull request, but I not sure on my edit in findDeepMarkerIndex (L.PM.Edit.Line.js), maybe it will be incorrect work in some cases and I haven`t time to think deeper.
Attach GIF how it works:
![ice_video_20190829-104532_edit_0(2)](https://user-images.githubusercontent.com/14214016/63921852-39f20b80-ca4c-11e9-8858-0eefdc441acc.gif)

**Comment by codeofsumit** [comment]
@lobovmaxim great! Thanks for sharing! I'm wondering do you think this could be application code only or did you have to do changes in leaflet.pm ?
If so, please create a PR just for inspiration (doesn't matter if it's merge-able). I can use it as a base then.

Thanks again for sharing, it looks great!

**Comment by lobovmaxim** [comment]
@codeofsumit I have forked and pull request. Glad to help you ;)

**Comment by lobovmaxim** [comment]
@codeofsumit Did you received the changes?

**Comment by sathiscode** [comment]
@lobovmaxim Thanks for the fix and its really helpful for my scenario. @codeofsumit , Please let us know on when we can see the changes integrated.

**Comment by AyoobMH** [comment]
@lobovmaxim Thank for this amazing fix, I'm currently facing this issue in my map, as I have too many points in each polygon. It would be amazing if these changes are integrated.

**Comment by spatselcci** [comment]
@codeofsumit would be great to know when/if changes like @lobovmaxim 's might be integrated. What's the best way to stay posted other than checking the commit history?

**Comment by codeofsumit** [comment]
Can’t give a timeline right now but on the Demo page is a mailing list sign up where I will regularly give updates to the development

**Comment by codeofsumit** [comment]
Alright people, it took me a while but I'm very close.
Take a look at this:

![perf](https://user-images.githubusercontent.com/2399810/76490923-9fc44480-642c-11ea-9ab8-79a950d264fd.gif)

I have some cool stuff coming up for you 🚀

**Comment by codeofsumit** [comment]
We are very very close. The next release will fly 🚀
![perf2](https://user-images.githubusercontent.com/2399810/76546867-84892180-648c-11ea-8214-952a33dced63.gif)
![perftest](https://user-images.githubusercontent.com/2399810/76546870-8521b800-648c-11ea-93a7-2d1b4c00484e.gif)
![perf](https://user-images.githubusercontent.com/2399810/76546854-7f2bd700-648c-11ea-9e90-8065f4a74991.gif)

**Comment by codeofsumit** [comment]
Alright, I'm basically done. I have just sent the mentioned improvements and 3 more options to Leaflet-Geoman Pro customers to test. When they come back with positive feedback, I will merge the feature branch #557 so this usecase here should be no problem anymore 🚀 

![image](https://user-images.githubusercontent.com/2399810/76707785-8c49ff80-66f2-11ea-8211-3eeea5ea3679.png)

**Comment by codeofsumit** [comment]
# **ITS DONE** 🎉 🎉 

It took me weeks and months to build, but I have [just released Leaflet-Geoman 2.5](https://github.com/geoman-io/leaflet-geoman/releases) with major performance improvements on large datasets. Adding data, switching edit-modes etc, will no longer drag down your user experience 🚀.

Additionally, this release includes a new option `limitMarkersToCount` which will improve performance even further when editing large layers.

![](https://user-images.githubusercontent.com/2399810/76546867-84892180-648c-11ea-8214-952a33dced63.gif)

### **There's more** ⭐️ 
For large professional projects, I have released *4 more options* in Leaflet-Geoman Pro ⭐️ to finetune the performance for your usecase.

- `limitMarkersCountGlobally`
- `limitMarkersToZoom`
- `limitMarkersToViewport`
- `limitMarkersToClick`

They all can be mixed and matched however you want, which will result in a snappy and smooth user experience even with massive amounts of data on the map. I have written [a detailed article](https://geoman.io/blog/performance-for-massive-geodata) going into each option.

![](https://blog.raum.sh/content/images/2020/03/perfoptionscombined.gif)

# Performance Discount
[Leaflet-Geoman Pro ](https://geoman.io/leaflet-geoman#pro)⭐️ is available now. To celebrate it's launch and support you in adopting it, I want to give you all a **50% discount on the Pro** library. Just use the discount code `performance` during checkout, or use [this direct link](https://gumroad.com/l/skMCA/performance).

I hope the newest releases of Leaflet-Geoman and Leaflet-Geoman Pro will make your projects more awesome. As always, let me know if I can help out in any way - I'm here and thanks to the Pro customers, more active than ever 🚀
